### PR TITLE
refactor: cleanup functions in api + worker

### LINKED ISSUE #190: graceful shutdown of worker and hatchet-api

### PR DESCRIPTION
**Pull request by steebchen**
# Description

Adds cleanup functions to the API and the worker instance instead of cancelling via context

## Type of change

- [x] Refactor

# What's Changed

- [x] cleanup functions in worker
- [x] proper cleanup in api
- [x] refactor api

### CODE REVIEW

File: pkg/client/dispatcher.go [hunk 1]
```diff
@@ -243,15 +243,14 @@ func (a *actionListenerImpl) Actions(ctx context.Context, errCh chan<- error) (<
 
 					if err != nil {
 						a.l.Error().Msgf("Failed to subscribe: %v", err)
-						errCh <- fmt.Errorf("failed to subscribe: %w", err)
-						return
+						panic(fmt.Errorf("failed to subscribe: %w", err))
```
Comment (steebchen) [inline pkg/client/dispatcher.go:246]: is this fine?

File: pkg/client/dispatcher.go [hunk 2]
```diff
@@ -228,7 +228,7 @@ func (a *actionListenerImpl) Actions(ctx context.Context, errCh chan<- error) (<
 
 					if err != nil {
 						a.l.Error().Msgf("Failed to close send: %v", err)
-						errCh <- fmt.Errorf("failed to close send: %w", err)
+						panic(fmt.Errorf("failed to close send: %w", err))
```
Comment (steebchen) [inline pkg/client/dispatcher.go:231]: is this fine?

File: pkg/client/dispatcher.go [hunk 3]
```diff
@@ -243,16 +243,14 @@ func (a *actionListenerImpl) Actions(ctx context.Context, errCh chan<- error) (<
...
+						panic(fmt.Errorf("failed to subscribe: %w", err))
 					}
 
 					continue
 				}
 
 				a.l.Error().Msgf("Failed to receive message: %v", err)
-				errCh <- fmt.Errorf("failed to receive message: %w", err)
-				return
+				panic(fmt.Errorf("failed to receive message: %w", err))
```
Comment (steebchen) [inline pkg/client/dispatcher.go:253]: is this fine?

File: pkg/client/dispatcher.go (hunk 2 above)
Comment (abelanger5) [reply pkg/client/dispatcher.go:231]: This one is fine

File: pkg/client/dispatcher.go (hunk 1 above)
Comment (abelanger5) [reply pkg/client/dispatcher.go:246]: Yes, this is after a retrier so makes sense.

File: pkg/client/dispatcher.go (hunk 3 above)
Comment (abelanger5) [reply pkg/client/dispatcher.go:253]: Yes

**Review by abelanger5 — APPROVED** [review]

**Pull request merged**